Implement on-type-formatting support

nemethf opened the pull request:
* eglot.el (eglot--on-type-formatting-chars): New local var, set
and unset ...
(eglot--managed-mode): ... here based on server's
documentOnTypeFormatting capability, ...
(eglot--post-self-insert-hook): and used here to call eglot-format
for on-type-formatting characters.
(eglot-format): Add new optional argument character to request
:textDocument/onTypeFormatting.

* eglot-tests.el (eglot--simulate-key-event): New helper defun.
(rust-on-type-formatting): New test.

joaotavora commented on eglot.el:
```diff
@@ -1974,7 +1989,9 @@ THINGS are either registrations or unregisterations (sic)."
 
 (defun eglot--post-self-insert-hook ()
   "Set `eglot--last-inserted-char'."
```
This docstring needs an overhaul, it should become something more generic like `Do Eglot things after a character is inserted.`, maybe with better imagination.

joaotavora commented on eglot.el:
```diff
@@ -2356,14 +2373,22 @@ Try to visit the target file for a richer summary line."
+                 `(:textDocument/onTypeFormatting
+                   :documentOnTypeFormattingProvider
+                   ,`(:position ,(eglot--pos-to-lsp-position beg)
+                      :ch ,character)))
```
You can put the `(string character)` here, or `(string last-input-event)`

joaotavora commented on eglot.el:
```diff
@@ -2356,14 +2373,22 @@ Try to visit the target file for a richer summary line."
-is not active."
+is not active.
+
+If CHARACTER is non-nil, assume it is an on-type-formatting,
```
Say the type of `CHARACTER`, i.e. that it is a character (or a one-character string).  Perhaps name it `ON-TYPE-FORMAT`.

joaotavora commented on eglot.el:
```diff
@@ -1651,7 +1665,8 @@ Use `eglot-managed-p' to determine if current buffer is managed.")
                    (null (eglot--managed-buffers server)))
-          (eglot-shutdown server))))))
+          (eglot-shutdown server))))
+    (setq eglot--on-type-formatting-chars nil)))
```
If this were a server slot you wouldn't need to reset it.

joaotavora commented on eglot.el:
```diff
@@ -1623,7 +1626,18 @@ Use `eglot-managed-p' to determine if current buffer is managed.")
     (unless (eglot--stay-out-of-p 'eldoc) (eldoc-mode 1))
-    (cl-pushnew (current-buffer) (eglot--managed-buffers (eglot-current-server))))
+    (cl-pushnew (current-buffer) (eglot--managed-buffers (eglot-current-server)))
+    (setq eglot--on-type-formatting-chars
```
I think this should be saved in a server slot somewhere inside the `:success-fn` callback of `eglot-connect`.

joaotavora commented on eglot.el:
```diff
@@ -1623,7 +1626,18 @@ Use `eglot-managed-p' to determine if current buffer is managed.")
+    (cl-pushnew (current-buffer) (eglot--managed-buffers (eglot-current-server)))
+    (setq eglot--on-type-formatting-chars
+          (cl-loop
+           for char in (append
```
I think you just just use `cons` here.  And maybe avoid the loop completely and just store the list of 1-char strings in the slot.  You can convert back to char when doing the test, I don't think it's expensive.

joaotavora commented on eglot.el:
```diff
@@ -1974,7 +1989,9 @@ THINGS are either registrations or unregisterations (sic)."
   "Set `eglot--last-inserted-char'."
-  (setq eglot--last-inserted-char last-input-event))
+  (setq eglot--last-inserted-char last-input-event)
+  (when (memq last-input-event eglot--on-type-formatting-chars)
```
I would use `cl-find` with a `:key` that extracts the char.  Putting more of my ideas together, this would be
```lisp
(when-let* ((server (eglot-current-server))
            (char (member (string last-input-event)
                          (eglot--on-type-formatting server))))
  (eglot-format beg nil char))
```

joaotavora commented on eglot-tests.el:
```diff
@@ -670,6 +677,29 @@ pyls prefers autopep over yafp, despite its README stating the contrary."
+          (should (eglot--tests-connect))
+          ;; Idle until the server finishes its housekeeping by
+          ;; waiting for publishDiagnostics.  This notification
+          ;; arrives quite late.  experimental/serverStatus arrives
```
Then maybe wait for `experimental/serverStatus`?  Is it too unstable, you think?

nemethf replied:
It is stable, but the server only sends it if the client requires it.  I implemented this [here](https://github.com/nemethf/eglot-x/blob/master/eglot-x.el#L189-L193), but I though that's too intrusive for eglot-test.el.  Shall I copy that here?

joaotavora commented on eglot-tests.el:
```diff
@@ -262,6 +262,13 @@ Pass TIMEOUT to `eglot--with-timeout'."
          (eglot-connect-timeout timeout))
     (apply #'eglot--connect (eglot--guess-contact))))
 
+(defun eglot--simulate-key-event (char)
```
There is `ert-simulate-keys` in `ert-x.el`.  A hack it itself seems less hacky than this (and not our responsibility ;-) )  Doesn't it work?  If it doesn't then fine to add this substitute.

nemethf replied:
`ert-simulate-keys` is not in Emacs 27.1, but more importantly it does something else: it provides user input for a function if it queries the user input (eg. read-from-minibuffer) during its execution.  The goal of `eglot--simulate-key-event` is initiate a command-loop for a given keypress.  I think the whole thing can be replaced with `(setq last-input-event ?.) (self-insert-command 1 ?.)` as that also calls post-self-insert-hooks.  But then some part of the implementation becomes part of the test.  An earlier iteration of this PR redefined keybindings for the shake of execution speed.  eglot--simulate-key-event worked with that implementation as well.

Shall I (a) change the test to use `(setq last-input-event ?.) (self-insert-command 1 ?.)`, or (b) extend the documentation of  `eglot--simulate-key-event`?

joaotavora commented on eglot.el:
```diff
@@ -1548,6 +1548,9 @@ and just return it.  PROMPT shouldn't end with a question mark."
 (defvar-local eglot--saved-bindings nil
   "Bindings saved by `eglot--setq-saving'.")
 
+(defvar-local eglot--on-type-formatting-chars nil
```
Make this a server slot. If nothing else, it'll simplify things when we support multiple servers.  _if_ we ever support that :-)

joaotavora reviewed (commented):
Thanks very much! Looks good, but please consider my comments and changes.

joaotavora commented:
Actually, I see that even the new server slot isn't needed.  This is already in the server as "capabilities".  So just grab it from there.  I don't think processing required is extremely more onerous.  At least not when compared to the performance hit we're already incurring in when calling `jsonrpc-request` from the `post-self-insert-hook`.

So this should simplify the code a lot.

nemethf commented:
> Actually, I see that even the new server slot isn't needed. This is already in the server as "capabilities". So just grab it from there. I don't think processing required is extremely more onerous. At least not when compared to the performance hit we're already incurring in when calling `jsonrpc-request` from the `post-self-insert-hook`.
> 
> So this should simplify the code a lot.

The goal of precomputing eglot--on-type-formatting-chars is make post-self-insert-hook run fast for chars that are not trigger chars for on-type-formatting.  Shall I (a) add this to the documentation of eglot--on-type-formatting-chars, or (b) simplify the code and calculate the list every time the user presses a key?

I'll address the rest of your comments once I see clear with these two questions.

Thanks for the review.

joaotavora commented:
Please use (b) and use cons to make list construction easier.

João


On Mon, Mar 28, 2022, 07:43 Felicián Németh ***@***.***>
wrote:

> Actually, I see that even the new server slot isn't needed. This is
> already in the server as "capabilities". So just grab it from there. I
> don't think processing required is extremely more onerous. At least not
> when compared to the performance hit we're already incurring in when
> calling jsonrpc-request from the post-self-insert-hook.
>
> So this should simplify the code a lot.
>
> The goal of precomputing eglot--on-type-formatting-chars is make
> post-self-insert-hook run fast for chars that are not trigger chars for
> on-type-formatting. Shall I (a) add this to the documentation of
> eglot--on-type-formatting-chars, or (b) simplify the code and calculate the
> list every time the user presses a key?
>

joaotavora commented on eglot.el:
```diff
@@ -1973,8 +1973,18 @@ THINGS are either registrations or unregisterations (sic)."
+  "Set `eglot--last-inserted-char', call on-type-formatting if necessary."
+  (setq eglot--last-inserted-char last-input-event)
+  (when (or (eq last-input-event
+                (aref (eglot--server-capable
```
won't this `aref` :boom: if the server is not capable of this?

nemethf commented:
I've force-pushed a new version addressing your concerns including the aref :boom:.

It turned out the experimental/serverStatus is just as slow a waiting for the first diagnostic notification.

joaotavora commented:
I took the liberty of making some minor tweaks and pushed this.  Let's keep eyes peeled for any problems (performance, :boom:, etc), though I think this is pretty sane.

nemethf commented:
Thanks!
